File: Scheduling/movie_schedule.js
const datetime = require('moment');
const FORMAT = "YYYY-MM-DD";
// let startDate = datetime("2019-05-11", "YYYY-M-DD HH:mm:ss")
// let endDate = datetime("2019-06-11", "YYYY-M-DD HH:mm:ss")
// let diff = endDate.diff(startDate, "days");
// // console.log(diff);
// let now = datetime();
// // console.log(now.format("YYYY-MM-DD"));

class DateRange {
    constructor(start, end) {
        this.start = start;
        this.end = end;
        this.difference = end.diff(start, "days");
        this.format = "YYYY-MM-DD";
    };

    printRange() {
        return `${this.start.format(this.format)} ===> ${this.end.format(this.format)}`;
    };

    printStart() {
        return `${this.start.format(this.format)}`;
    };

    printEnd() {
        return `${this.end.format(this.format)}`;
    };
    
    greaterRange(dr2) {
        return this.difference > dr2.difference;
    };

    equalRange(dr2) {
        return this.difference == dr2.difference;
    };

    startsFirst(dr2) {
        return this.start < dr2.start;
    };

    endsFirst(dr2) {
        return this.start < dr2.start;
    };

    overlap(dr2) {
        let latestStart = datetime.max(this.start, dr2.start);
        let earliestEnd = datetime.min(this.end, dr2.end);
        let delta = earliestEnd.diff(latestStart, "days") + 1;
        return Math.max(0, delta);
    };

    isOverlap(dr2) {
        return this.overlap(dr2) != 0;
    };

};

let makeDateTimeObject = function(day, month, year) {
    return datetime(`${year}-${month}-${day}`, FORMAT);
};

let getRandomInteger = function(min, max) {
    return Math.floor(Math.random() * (max-min)) + min;
};

let getRandomDate = function(start) {
    const MAX_DAYS = 30;
    const MIN_DAYS = 3;
    let temp = start.clone();
    let numberOfDays = getRandomInteger(MIN_DAYS, MAX_DAYS);
    numberOfDays = numberOfDays%temp.daysInMonth() == numberOfDays ? 
                   numberOfDays : temp.daysInMonth() - numberOfDays;
    let newDate = datetime(temp.add(numberOfDays, "days").format(FORMAT));
    return newDate;
};

let getDate = function(spread, year) {
    let month = getRandomInteger(1,spread);
    let _temp = datetime(`${year}-${spread}`, FORMAT);
    let day = getRandomInteger(1, _temp.daysInMonth());
    return makeDateTimeObject(day, month, year);
};

let generateJobs = function(n, spread, year) {
    let dateRanges = [];
    for(let i=0; i < n; i++) {
        let start = getDate(spread, year);
        let end = getRandomDate(start);
        dateRange = new DateRange(start, end);
        dateRanges.push(dateRange);
    };
    return dateRanges;
};

let any = function(iterable) {
    for (let idx=0; idx<iterable.length; idx++) {
        if(iterable[idx]){
            return true;
        }
    }
    return false;
};

let argsort = function(array, dec=false) {
    return Array.from(Array(array.length).keys()).sort((a,b) => 
           array[a] < array[b] ? (dec ? 0 : -1) : 
           (array[b] < array[a]) | (dec ? -1 : 0));
};

let checkOverlap = function(date, dateRanges) {
    let d;
    for (d of dateRanges) {
        console.log(`Checking overlap with previous date: ${d.printRange()}`);
        if (date.isOverlap(d)) {
            console.log(`Overlap with date: ${d.printRange()};`);
            return false;
        }
        else {
            console.log(`No overlap with date: ${d.printRange()}`);
            continue;
        }
    }
    console.log(`No overlap with any previous dates.`);

    return true;
};

let earliestJobFirst = function(dateRanges) {
    let datesToTake = [];
    let sortedDateRanges = argsort(dateRanges, true);
    datesToTake.push(dateRanges[sortedDateRanges[0]]);
    console.log(`Job which starts the first: ${dateRanges[sortedDateRanges[0]].printRange()}`);
    let otherDate;
    let takeDate;
    for(otherDate of sortedDateRanges.slice(1,)){
        console.log(`Next earliest starting job: ${dateRanges[otherDate].printRange()}`);
        takeDate = checkOverlap(dateRanges[otherDate], datesToTake);
    if(takeDate) {
        datesToTake.push(dateRanges[otherDate]);
        console.log(`Appending to taken jobs list: ${dateRanges[otherDate].printRange()}`);
    }
    console.log("=============================================");
    };
    return datesToTake;
};   


let x1 = datetime("2018-10-12", FORMAT);
let y1 = datetime("2019-11-12", FORMAT);

let x2 = datetime("2019-10-20", FORMAT);
let y2 = datetime("2019-11-20", FORMAT);

let X = new DateRange(x1, y1);
let Y = new DateRange(x2, y2);

console.log(x1.daysInMonth());

let dateRanges = generateJobs(30,7,2019);

let jobsToTake = earliestJobFirst(dateRanges);
for(let j=0;j<jobsToTake.length;j++){
    console.log(jobsToTake[j].printRange());
};

File: Scheduling/movieSchedule.java
import java.util.Date;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.text.ParseException;
import java.time.YearMonth;

class DateRange extends Date
{
    //instance fields
    private Date start;
    private Date end;
    private long length;
    private DateFormat sdf;

    //constructor
    public DateRange(Date x, Date y) {
        this.start = x;
        this.end = y;
        this.length = (this.end.getTime() - this.start.getTime())/(1000*60*60*24);
        this.sdf = new SimpleDateFormat("yyyy-MM-dd");
    }


    //methods
    public int getRange() {
        return this.length;
    }

    public boolean greaterThan(DateRange other) {
        return this.difference > other.difference;
    }

    public boolean equalTo(DateRange other) {
        return this.difference == other.difference;
    }

    public String printRange() {
        return this.sdf.format(this.start) + " ===> " + this.sdf.format(this.end);
        //return this.start.toString() + " ===> " + this.end.toString();
    }

    public int overlap(DateRange other) {
        Date latestStart = this.start.after(other.start) ? other : this.start;
        Date earliestEnd = this.end.before(other.end) ? this.end : other;
        int delta = (earliestEnd.getTime() - latestStart.getTime())/(1000*60*60*24);
        return Math.max(0, delta);
    }

    public boolean isOverlap(DateRange other) {
        return this.overlap(other) != 0;
    }

}

class Tests
{
    //instance fields
    private DateFormat sdf;
    //constructor
    public Tests()
    {
        this.sdf = new SimpleDateFormat("yyyy-MM-dd");
    }

    public int getRandomInteger(int min, int max) {
        return Math.floor(Math.random() * (max-min)) + min;
    }

    public getRandomDate(Date start) {
        const MAX_DAYS = 30;
        const MIN_DAYS = 3;

        Date temp = start.clone();
        int numberOfDays = this.getRandomInteger(MIN_DAYS, MAX_DAYS);
        YearMonth yearMonthObject = YearMonth.of(temp.getYear(), temp.getMonth());
        int daysInMonth = yearMonthObject.lengthOfMonth();
        numberOfDays = numberOfDays%daysInMonth == numberOfDays ?
                       numberOfDays : daysInMonth - numberOfDays;
        Date newDate = sdf.parse(temp.getTime() + numberOfDays*(1000*60*60*24));
        return newDate;
    }

}

public class movieSchedule {

    public long difference(Date x, Date y) {
        return (y.getTime() - x.getTime())/(1000*60*60*24);
    }

    public static void main(String[] args) throws ParseException {
        DateFormat df = new SimpleDateFormat("yyyy-MM-dd");
        Date today = new Date();
        System.out.println(df.format(today));
        String customDate = "2019-11-23";
        Date newDate = df.parse(customDate);
        System.out.println(newDate.compareTo(today));

        System.out.println(newDate.after(today));
        movieSchedule ms = new movieSchedule();
        long dd = ms.difference(today, newDate);
        System.out.println(dd);

        DateRange X = new DateRange(today, newDate);
        System.out.println(X.getDifference());
        System.out.println(X.printRange());

    }
}
